swarm/src/connection.rs

```diff
@@ -1008,6 +1004,88 @@ mod tests {
         assert!(start.checked_add(duration).is_some())
     }
 
+    #[test]
+    fn compute_new_shutdown_does_not_panic() {
+        let _ = env_logger::try_init();
+
+        #[derive(Clone, Debug)]
+        struct ArbitraryKeepAlive(KeepAlive);
+
+        impl Arbitrary for ArbitraryKeepAlive {
+            fn arbitrary(g: &mut Gen) -> Self {
+                let keep_alive = match g.gen_range(1u8..4) {
+                    1 => KeepAlive::Until(
+                        Instant::now()
+                            .checked_add(Duration::from_secs(u64::arbitrary(g)))
+                            .unwrap_or(Instant::now()),
+                    ),
+                    2 => KeepAlive::Yes,
+                    3 => KeepAlive::No,
+                    _ => unreachable!(),
+                };
+
+                Self(keep_alive)
+            }
+        }
+
+        #[derive(Debug)]
+        struct ArbitraryShutdown(Shutdown);
+
+        impl Clone for ArbitraryShutdown {
+            fn clone(&self) -> Self {
+                let shutdown = match self.0 {
+                    Shutdown::None => Shutdown::None,
+                    Shutdown::Asap => Shutdown::Asap,
+                    Shutdown::Later(_, instant) => Shutdown::Later(
+                        // compute_new_shutdown does not touch the delay. Delay does not
+                        // implement Clone. Thus use a placeholder delay.
+                        Delay::new(Duration::from_secs(1)),
+                        instant.clone(),
+                    ),
+                };
+
+                ArbitraryShutdown(shutdown)
+            }
+        }
+
+        impl Arbitrary for ArbitraryShutdown {
```

Comment:
Same here.

Reply:
Oh interesting, thanks for clarifying!

Reply:
Can we panic on the delay cloning instead?

Reply:
`quickcheck` uses `Clone` to run through the various permutations, e.g. change one parameter but not another. For the latter it needs `Clone`. Thus a panic in a `Clone` implementation would not work.